pip/pip-443.md

```diff
@@ -0,0 +1,233 @@
+
+Without following the move-before-publish pattern, we have to make `recycle` synchronized, which will have more performance overhead:
+
+```java
+    public synchronized void recycle() {
+        if (this.object != null) {
+            this.object = null;
+            handle.recycle(this);
+        }
+    }
+```
+
+## Issue 3: questionable benefits
+
+How much GC pressure can be reduced by sacrificing the memory allocation latency is hard to measure and never proved by any benchmark. When `Recycler` was introduced in Netty, Java 8 was even not released. The Java GC has involved a lot since these years. The main use case of `Recycler` in Netty is for `ByteBuf` to avoid the time-consuming buffer memory allocation from direct memory, where the buffer is used as a request for users to process in the `FastThreadLocalThread`.
+
+Even if assuming a usage of `Recycler` can reduce the GC pressure significantly in a scenario, the benefit can be easily broken by a slight change of the code, which is hard to be noticed and verified.
+
+In Pulsar, Netty `Recycler` is typically used to wrap a list of objects to avoid passing them as method parameters, which makes methods too long with duplicated parameters. For example, [`ManagedLedgerImpl#asyncReadOpEntry`](https://github.com/apache/pulsar/blob/d833b8bef21cb9e85f0f313eb9d49c7ca550fbbd/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L2064) accepts an `OpReadEntry` instance, which is recycable and includes many fields. This method is called for each read operation, which should happen frequently, so technically it can save heap memory allocations.
+
+However, as the code becomes complicated, after being passed to the entry cache to process, the `OpReadEntry` is wrapped as another callback object: https://github.com/apache/pulsar/blob/d833b8bef21cb9e85f0f313eb9d49c7ca550fbbd/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/RangeEntryCacheImpl.java#L349-L365
+
+Now, if the recycler design on `OpReadEntry` saved 1 of 1 heap memory allocation for each read operation, now it just saves 1 of 2 heap memory allocations. What's more, this wrapped callback is passed to `PendingReadsManager` and eventually used as a field of a new record type `ReadEntriesCallbackWithContext` to an `ArrayList` in https://github.com/apache/pulsar/blob/d833b8bef21cb9e85f0f313eb9d49c7ca550fbbd/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/cache/PendingReadsManager.java#L256
+
+Now, each read operation has:
+- 1 object allocated by Netty recycler (`OpReadEntry`)
+- 1 object allocated from heap memory (the wrapped callback in `RangeEntryCacheImpl`)
+- 1 object allocated from heap memory (the `ReadEntriesCallbackWithContext` object that has a field of the wrapped callback)
+  - This object is added to an `ArrayList` allocated from heap memory with at least 1 element
+
+Eventually, only saving the 1st heap memory allocation should not make a difference.
```

Comment:
btw. I guess allocations in the JVM aren't only about the heap allocations. If JVM's JIT/hotspot escape analysis determines an object doesn't escape its scope, the JVM can optimize it away through scalar replacement or stack allocation, eliminating the need for GC completely. Such optimizations in the JVM make it even more justified to get rid of the recycler.

Reply:
Yes, escape analysis should be here even with an older version of GC, especially for short-lived objects.

I've made some minor adjustments in the latest commit.